security: Add how to revoke a key for imx (#318)
foundriesio/docs: source/reference-manual/security/revoke-imx-keys.rst

.. highlight:: sh

.. _ref-revoke-imx-keys:

Revoke Boot keys on IMX
=======================

This tutorial describes the step-by-step on how to revoke a SRK key on i.MX SoC boards, the purpose of revocations and where to find more detail on this topic.

What is Revocation of a Key
---------------------------

The process to revoke a SRK key is used when one or more SRK keys become compromised.

> When a SRK key is revoked, that key is considered not feasible, and any image signed with that key is not executed during boot. This is a permanent change to

Author: @vanmaegima I struggled to find a word for describing what happens to the key as I'm not into the security language. So thanks for your suggestion! I think it's better than the first one.

the hardware.

In this tutorial, a root of trust with 4 SRK keys fused to the i.MX machine is used as example. For a different architecture, some commands may differ.
Details on how to setup a secure boot on i.MX machines on

> The following table lists the values for each SRK key. The value for ``SRK_REVOKE`` mask is used for the revocation command line and is detailed on next sections.

Author: I add a line describing it. could you please give a look?

> The value for ``srk-index`` is a parameter for the ``sign-file.sh`` script used to sign the artifacts used during the boot.

Reviewer: this suggestion was to avoid the word "used" twice

.. _srk-revoke-imx:

.. csv-table:: **SRK Revoke i.MX**
   :file: ../../_static/csv/revoke_imx.csv
   :widths: 30, 30, 30, 30, 30
   :header-rows: 1

> ``srk-index``: this is a ``sign-file.sh`` parameter used to define a SRK key to be used.

Reviewer: same here

Columns 1, 2 and 3 are from Table 4 from  `Secure Boot Using HABv4 Guide`_.

.. note::

>     On a HAB environment, only the first 3 SRK keys can be revoked. On a AHAB environment, all 4 SRK can be revoked `Code-Signing Tool User's Guide`_.

Reviewer: ```suggestion
    On a HABv4 environment, only the first 3 SRK keys can be revoked. On a AHAB environment, all 4 SRK can be revoked. For more details, check `Code-Signing Tool User's Guide`_.
```

    When there is no available SRK key, the board cannot boot!

> How to sign the boot image
> --------------------------

Reviewer: ```suggestion
How to sign the boot image for revoking a key
---------------------------------------------
```

> The first step is to make sure there are other SRK key available for the boot. The revocation can only be executed after a secure boot is executed, with the unlock of ``SRK_REVOKE`` write access. After that a fuse is burned, in short:

Reviewer: ```suggestion
The first step is to make sure there are other SRK keys available for the boot. A key can only be revoked after a secure boot with a different key is executed, with the permission to unlock ``SRK_REVOKE`` write access. After that a fuse is burned, in short:
```
this is what I understand from our previous conversations, is this correct?

>     * boot with an available SRK key

Reviewer: ```suggestion
    * boot with an available SRK key, different from the one to revoke
```

    * unlock ``SRK_REVOKE``
    * fuse a register accordingly to the SRK key being revoked.

> The sign process is based on the commands  from, adding two parameters:

Reviewer: ```suggestion
The sign process is based on the commands from ..., adding two parameters:
```
missing link here, I guess

Author: I add the link

>     ``--srk-index = <value>`` to chose the SRK key to be used on that boot sequence. For the right value, consult :ref:`srk-revoke-imx`.

Reviewer: ```suggestion
    ``--srk-index = <value>`` to choose the SRK key to be used on that boot sequence. For the right value, consult :ref:`srk-revoke-imx`.
```

    ``--enable-revoke`` to unlock the key revocation register write access.

For example, for ``imx6ullevk-sec`` SPL image to be signed with SRK1, use the following command::

    #Sign the MFGTool SPL file
    ./sign-file.sh --engine SW --key-dir $KEY_PATH \
                --cst $CST_PATH \
                --spl $SPL_PATH/mfgtool-files-imx6ullevk-sec/SPL-mfgtool \
                --fix-sdp-dcd \
                --srk-index 2 \
                --enable-revoke

    #Sign the SPL file
    ./sign-file.sh --engine SW --key-dir $KEY_PATH \
                --cst $CST_PATH \
                --spl $SPL_PATH/SPL-imx6ullevk-sec \
                --srk-index 2

> In the example, only the SPL from MFGtool has the write access permission to revoke the key, as the suggestion is to use ``bootloader.uuu`` script to load the MFGTool SPL and U-Boot prompt to perform the fuse programing command. For other boot scripts, it might be required to include ``--enable-revoke`` to the SPL file signing process (second command).

Reviewer: ```suggestion
In the example, only the SPL from MFGtool has the write access permission to revoke the key, as the suggestion is to use ``bootloader.uuu`` script to load the MFGTool SPL and U-Boot prompt to perform the fuse programming command. For other boot scripts, it might be required to include ``--enable-revoke`` to the SPL file signing process (second command).
```

Author: I personally prefer to not address roadmap statements on the documentation because it's tough to maintain once it's published.

How to burn the fuse
--------------------

The suggestion is to use the ``bootloader.uuu`` script to access U-Boot prompt to execute the following command::

    fuse prog <bank> <word> <hexval>

The values for ``<bank>`` and ``<word>`` for the register ``SRK_REVOKE`` can be found on the SoC Reference Manual. The value for ``<hexval>`` is from :ref:`srk-revoke-imx`.

> 

Reviewer: I feel like we should have a warning here mentioning again that after revoking a key we cannot boot with it again, that we need at least 1 srk available to continuing booting that board, that it's an irreversible process, etc...

Author: I add a warning. Please give a look

For example, for ``imx6ullevk-sec``, to revoke SRK2::

    fuse prog 5 7 0x4
    Programming bank 5 word 0x00000007 to 0x00000004...
    Warning: Programming fuses is an irreversible operation!
            This may brick your system.
            Use this command only if you are sure of what you are doing!

    Really perform this fuse programming? <y/N>
    y

> The following error happens when the key revocation write access is not available and can be fixed by adding ``--enable-revoke`` during the sign of boot image::

Reviewer: ```suggestion
The following error happens when the key revocation write access is not available (SRK_REVOKE is not unlocked) and can be fixed by adding ``--enable-revoke`` during the sign of boot image::
```

Reviewer: maybe this is a note

For example, for ``imx6ullevk-sec``, to revoke SRK2::

    fuse prog 5 7 0x4
    Programming bank 5 word 0x00000007 to 0x00000004...
    Warning: Programming fuses is an irreversible operation!
            This may brick your system.
            Use this command only if you are sure of what you are doing!

    Really perform this fuse programming? <y/N>
    y

The following error happens when the key revocation write access is not available (``SRK_REVOKE`` is not unlocked) and can be fixed by adding ``--enable-revoke`` during the sign of boot image::

    mxc_ocotp fuse_prog(): Access protect error
    ERROR

After the revocation of SRK2, it cannot be used to boot that board never again. Test it by signing again using this SRK key and the boot must fail.

How to Revoke a Key for Devices on a Fleet
------------------------------------------

The method suggested here describes the commands needed to revoke a key from the SoC perspective. It depends on serial download, console and bootloader access, which are not always accessible on devices on the field. However, this is the base procedure to be used on a fleet.

> The automatization can take place on FoundriesFactory level by creating a signed SPL using another SRK key and enabling the ``SRK_REVOKE`` write access. While on this, the ``bootcmd`` can be customized to perform the fusing command needed to revoke the compromised key.

Reviewer 2: I'm unsure if automatization is the correct word, perhaps authorization or authentication?  Though I don't fully understand the context, so automatization may be correct, feel free to ignore!

Author: I have updated the sentence to try to make it more clear. Could you, please, take a look @kprosise ?

Reviewer 2: Yep!

The fusing can be performed on Linux Kernel mode instead, when the system is configured to allow this kind of execution.

Then the firmware update is performed in a Wave described in detail on :ref:`ref-production-targets`.

After the revocation wave, another firmware update wave is required, now with the bootloader configured to disable the write access to the ``SRK_REVOKE``, and still using the reliable SRK key.

> This is a two-steps process which is highly dependent on the device configuration and access, and requires caution. The revoke fusing command can make the device unavailable if not executed properly.

Reviewer 3: This looks like a statement that says the implementation is up to the user.

I guess I am okay with this.
